### DISCUSSION TITLE
Unnamed Subregisters in Pcode. How to find Base Register?

### DISCUSSION THREAD

**Discussion by wntress** [NationalSecurityAgency/ghidra#5308] [Q&A]
Architecture: AARCH64

``` 
002e6360 08 40 20 1e     fmov       s8,s0
                                                      s8 = COPY s0
                                                      *[register]0x5104:4 = COPY 0:4
                                                      *[register]0x5108:8 = COPY 0:8
                                                      *[register]0x5110:8 = COPY 0:8
                                                      *[register]0x5118:8 = COPY 0:8
```

There are several unnamed `*[register]offset:size` references above. This is causing problems for a project I'm working on, because I can't easily convert these register references into the owning parent/base register. Unnamed registers seem to happen a lot when working with the floating point / SIMD registers on AARCH64.

These references actually occur **inside** the parent register s8. The references are unnamed, because the offsets 0x5104-0x5118 do not correspond to a defined register.  Is there a function in the API that allows you to retrieve the parent/base register in this case?

VarnodeTranslator.getRegister() in the API fails to convert the register varnodes from Pcode to Register objects. This means you cannot use Register.getParentRegister() or Register.getBaseRegister().  

It seems like the aforementioned functions only use offset and size to fetch the corresponding register. This means any subregister reference that starts at an offset for which no register is explicitly defined causes problems. 

To circumvent the issue, you can write a function that searches to the left of the given offset until it finds a defined parent register (s8 in this case). Has this functionality already been implemented somewhere?

From the AARCH64 sleigh spec:
```
define register offset=0x5000 size=4
[
	s0 _ _ _ _ _ _ _
	s1 _ _ _ _ _ _ _
	s2 _ _ _ _ _ _ _
	s3 _ _ _ _ _ _ _
	s4 _ _ _ _ _ _ _
	s5 _ _ _ _ _ _ _
	s6 _ _ _ _ _ _ _
	s7 _ _ _ _ _ _ _
	s8 _ _ _ _ _ _ _
	...
```

Register definitions are here: https://github.com/NationalSecurityAgency/ghidra/blob/480f1e3a8bed20854bf5c2ac68f2ee48a33db966/Ghidra/Processors/AARCH64/data/languages/AARCH64instructions.sinc

**Comment by mumbel** [comment]
I thought maybe the `Language` class would be the answer, but I was unable to get any results from any of the relevant looking functions. 
 Was trying something along the lines of this but never got any register.  You might be able to parse the AddressSetView from `lang.getRegisterAddresses()`  

```python
from ghidra.program.model.address import AddressSpace
lang = currentProgram.getLanguage()
reg = lang.getRegister(AddressSpace.DEFAULT_REGISTER_SPACE, 0x5000, 4)
```

**Reply by wntress** [reply]
Parsing `lang.getRegisterAddresses()` looks like the most promising approach so far. Thank you
